src/evm.rs

```diff
@@ -1,179 +1,312 @@
-use crate::{
-    types::{AccountInfo, Env},
-    utils::addr,
-};
-use foundry_evm::{
-    backend::Backend,
-    executors::{Executor, ExecutorBuilder},
-    fork::CreateFork,
-    opts::EvmOpts,
-    utils::RuntimeOrHandle,
-};
-use pyo3::exceptions::PyRuntimeError;
-use pyo3::prelude::*;
-use revm::{primitives::U256, Database};
+use std::collections::HashMap;
 use std::fmt::Debug;
+use std::sync::Arc;
+use ethers_core::types::BlockId;
+use std::str::FromStr;
+
+use pyo3::prelude::*;
+use revm::{Database, DatabaseCommit, DatabaseRef, Evm, primitives::U256};
+use revm::db::{CacheDB, EthersDB};
+use revm::precompile::{Address, Bytes};
+use revm::precompile::B256;
+use revm::primitives::{AccountInfo as RevmAccountInfo, BlockEnv, CreateScheme, Env as RevmEnv, EnvWithHandlerCfg, ExecutionResult, HandlerCfg, Output, ResultAndState, SpecId, State, TransactTo, TxEnv};
+use revm::primitives::ExecutionResult::Success;
+use tracing::{trace, warn};
+use ethers_providers::{Http, Provider};
+
+use crate::{types::{AccountInfo, Env}, utils, utils::addr};
+use crate::empty_db_wrapper::EmptyDBWrapper;
+use crate::utils::{pydict, pyerr};
+
+type MemDB = CacheDB<EmptyDBWrapper>;
+type ForkDB = CacheDB<EthersDB<Provider<Http>>>;
+#[derive(Clone, Debug)]
+pub enum DB<'a> {
+    Memory(&'a MemDB),
+    Fork(&'a ForkDB),
+}
 
+
+#[derive(Clone, Debug)]
 #[pyclass]
-pub struct EVM(Executor);
+pub struct EVM {
+    /// The underlying `Database` that contains the EVM storage.
+    pub db: MemDB,
+    /// The EVM environment.
+    pub env: RevmEnv,
+    /// the current handler configuration
+    pub handler_cfg: HandlerCfg,
+    /// The gas limit for calls and deployments. This is different from the gas limit imposed by
+    /// the passed in environment, as those limits are used by the EVM for certain opcodes like
+    /// `gaslimit`.
+    gas_limit: U256,
 
-impl EVM {
-    pub fn db(&self) -> &Backend {
-        &self.0.backend
-    }
+    /// The fork to use for the EVM. If `None`, the EVM will use the memory database.
+    fork: Option<ForkDB>,
 }
 
-fn pyerr<T: Debug>(err: T) -> pyo3::PyErr {
-    PyRuntimeError::new_err(format!("{:?}", err))
+impl EVM {
+    pub fn db(&self) -> DB<'_> {
```

Comment:
Is there a reason `DB` enum is not stored as a field?

Reply:
It's just how the code evolved, I'll move it